File: examples/typescript/real-world.ts
// examples/typescript/real-world.ts
// A file search tool demonstrating progress, logging, and cancellation.
// Run: pmcp dev examples/typescript/real-world.ts

import { tool, ToolResult, ToolContext, ServerLogger } from 'protomcp';
import { z } from 'zod';
import * as fs from 'fs';
import * as path from 'path';

// Note: ServerLogger requires a transport send function. In a real pmcp process,
// this is wired up automatically by the runner. For demonstration purposes,
// we show the API shape — logging calls are forwarded to the MCP host.

tool({
  name: 'search_files',
  description: 'Search files in a directory by glob pattern',
  readOnlyHint: true,
  args: z.object({
    directory: z.string(),
    pattern: z.string(),
    max_results: z.number().default(50),
  }),
  handler({ directory, pattern, max_results }, ctx: ToolContext) {
    if (!fs.existsSync(directory)) {
      return new ToolResult({
        result: `Directory not found: ${directory}`,
        isError: true,
        errorCode: 'INVALID_PATH',
        message: 'The specified directory does not exist',
        suggestion: 'Check the path and try again',
      });
    }

    const matches: string[] = [];
    const allFiles: string[] = [];

    function walk(dir: string) {
      for (const entry of fs.readdirSync(dir, { withFileTypes: true })) {
        const full = path.join(dir, entry.name);
        if (entry.isDirectory()) walk(full);
        else allFiles.push(full);
      }
    }
    walk(directory);

    const total = allFiles.length;

    for (let i = 0; i < total; i++) {
      if (ctx.isCancelled()) {
        return new ToolResult({
          result: `Cancelled after scanning ${i}/${total} files. Found ${matches.length} matches so far.`,
          isError: true,
          errorCode: 'CANCELLED',
          retryable: true,
        });
      }

      if (i % 100 === 0) {
        ctx.reportProgress(i, total, `Scanning... ${i}/${total}`);
      }

      if (matchGlob(path.basename(allFiles[i]), pattern)) {
        matches.push(allFiles[i]);
        if (matches.length >= max_results) break;
      }
    }

    ctx.reportProgress(total, total, 'Complete');
    return new ToolResult({
      result: matches.length > 0 ? matches.join('\n') : 'No files found',
    });
  },
});

function matchGlob(filename: string, pattern: string): boolean {
  const regex = new RegExp(
    '^' + pattern.replace(/\*/g, '.*').replace(/\?/g, '.') + '$'
  );
  return regex.test(filename);
}

File: examples/python/real_world.py
# examples/python/real_world.py
# A file search tool demonstrating progress, logging, and cancellation.
# Run: pmcp dev examples/python/real_world.py

import os
import fnmatch
from protomcp import tool, ToolResult, ToolContext, log

@tool("Search files in a directory by glob pattern", read_only=True)
def search_files(ctx: ToolContext, directory: str, pattern: str, max_results: int = 50) -> ToolResult:
    log.info(f"Searching {directory} for '{pattern}'")

    if not os.path.isdir(directory):
        return ToolResult(
            result=f"Directory not found: {directory}",
            is_error=True,
            error_code="INVALID_PATH",
            message="The specified directory does not exist",
            suggestion="Check the path and try again",
        )

    matches = []
    all_files = []
    for root, dirs, files in os.walk(directory):
        for f in files:
            all_files.append(os.path.join(root, f))

    total = len(all_files)
    log.debug(f"Found {total} files to scan")

    for i, filepath in enumerate(all_files):
        if ctx.is_cancelled():
            log.warning("Search cancelled by client")
            return ToolResult(
                result=f"Cancelled after scanning {i}/{total} files. Found {len(matches)} matches so far.",
                is_error=True,
                error_code="CANCELLED",
                retryable=True,
            )

        if i % 100 == 0:
            ctx.report_progress(i, total, f"Scanning... {i}/{total}")

        if fnmatch.fnmatch(os.path.basename(filepath), pattern):
            matches.append(filepath)
            if len(matches) >= max_results:
                log.info(f"Hit max_results={max_results}, stopping early")
                break

    ctx.report_progress(total, total, "Complete")
    log.info(f"Search complete: {len(matches)} matches")
    return ToolResult(result="\n".join(matches) if matches else "No files found")


if __name__ == "__main__":
    from protomcp.runner import run
    run()
